ras-server.c

```diff
@@ -0,0 +1,363 @@
+
+static void ras_server_cleanup(void* arg) {
+	// Wait until main thread finishes current broadcast
+	pthread_mutex_lock(&server.fd_mutex);
+	server.nclients = 0;  // Prevent future broadcasts
+	pthread_mutex_unlock(&server.fd_mutex);
+	// De-initialize the remainder of the server attributes
+	pthread_mutex_destroy(&server.fd_mutex);
+	if(server.socketfd > 0) {
+		shutdown(server.socketfd, SHUT_RDWR);
+		close(server.socketfd);
+	}
+	if(server.fds)
+		free(server.fds);
+	memset(&server, 0, sizeof(struct ras_server));
+	signal(SIGPIPE, SIG_DFL);
+	log(ALL, LOG_INFO, "RAS server has stoped\n");
+}
+
+static void cleanup_unlock_mutex(void *p)
+{
+    pthread_mutex_unlock(p);
+}
+
+static void server_loop(void) {
+	int next_slot, server_full, rv, i;
+
+	pthread_cleanup_push(ras_server_cleanup, NULL);
+	server_full = 0;
+	next_slot = 0;
+
+	while(1) {
+		rv = poll(server.fds, server.nclients+1, -1);
+		if(rv < 0) {
+			log(ALL, LOG_ERR, "Can't poll the connection file descriptors\n");
+			break;
+		}
+		
+		pthread_mutex_lock(&server.fd_mutex);
+		pthread_cleanup_push(cleanup_unlock_mutex, &server.fd_mutex);
+		for(i = 0; i <= server.nclients; ++i) {
+			struct pollfd *p = &server.fds[i];
+			int clifd = 0;
+			
+			// Check if all poll() notifications have been served
+			if(!rv)
+				break;
+			// If a connection closed, release resources
+			if(p->revents & POLLHUP) {
+				--rv;
+				close(p->fd);
+				p->fd = -1;    // stop tracking in poll()
+				next_slot = i; // This slot is free
+				// Poll the socket if it there are connection slots
+				if(server_full) {
+					server_full = 0;
+					server.fds[server.nclients].fd = server.socketfd;
+					server.fds[server.nclients].events = POLLIN;
+				}
+				log(ALL, LOG_INFO, "Client %d disconnected from RAS server\n",
+					next_slot);
+			}
+			// If a connection opened, set-up context and invoke the handler
+			else if(p->revents & POLLIN) {
+				--rv;
+				// Find the next connection slot
+				if(next_slot < 0) {
+					for(int w = 0; w < server.nclients; ++w) {
+						if(server.fds[w].fd < 0) {
+							next_slot = w;
+							break;
+						}
+					}
+				}
+				clifd = accept(server.socketfd, NULL, NULL);
+				if(clifd < 0) {
+					log(ALL, LOG_WARNING, "Can't accept client connection\n");
+					continue;
+				}
+				// No more connection slots, stop polling the server socket
+				if(next_slot < 0) {
+					server.fds[server.nclients].fd = -1;
+					server_full = 1;
+					close(clifd);
+				} else {
+					server.fds[next_slot].fd = clifd;
+					next_slot = -1;
+					log(ALL, LOG_INFO, "Client %d connected to RAS server\n",
+						next_slot);
+				}
+			}
+		}
+		pthread_cleanup_pop(1); // Call pthread_mutex_unlock
+	}
+	pthread_cleanup_pop(1); // Call ras_server_cleanup
+}
+
+void ras_server_stop(void) {
+	if(server.tid > 0)
+		pthread_cancel(server.tid);
+}
+
+int ras_server_start(void) {
+	struct sockaddr_un addr;
+
+	server.socketfd = socket(AF_UNIX, SOCK_STREAM, 0);
```

Comment:
Maybe this could be made a little more generic, also allowing reporting issues to other machines via TCP/IP.

Reply:
Hi @mchehab, a remote report was one of my motivations for this functionality.
However, I chose to keep this out of RASdaemon as I thought it would not fit RASdaemon scope.

I envision that a local server feature is enough to enable a wide range of agents for handling the RAS event stream.
These agents can perform event pre-processing, reliability introspective analysis, and others without being tightly-coupled with RASdaemon.
In my opinion, the remote report feature is best suited for other software pieces that can determine their own protocol and/or the preferred system bus to do so.

What do you think?